lib/anagrams.rb

```diff
@@ -0,0 +1,44 @@
+class Anagrams
+  def initialize(dictionary)
+    @dictionary = dictionary
+  end
+
+  def group_anagrams
+    anagram_group = {}
+    read_file.group_by do |word|
+      sorted_word = sort_and_validate(word)
+      anagram_group[sorted_word] ||= []
+      (anagram_group[sorted_word] << word)
+    end
+    anagram_group.values
+  end
+
+  def find_anagram(input_word)
+    result = []
+    word_characters = sort_and_validate(input_word)
+    read_file.each do |word|
+      result.push(word) if sort_and_validate(word) == word_characters
```

Comment:
@zzokki81 whenever `find_anagram` is called you process the whole dictionary and check each word separately. That means for 1 word you will load each dictionary word once, process it, and then compare it. For the dictionary of 1000 words, you need to do this 1001 times (+1 is for a given word). For 50 words it'll be 1000 * 50 + 1 = 50001 processed words.

```
$ wc -l anagrams-wordlist.txt
  338882 anagrams-wordlist.txt
```

This means there are 338882 lines in that file: 338,882*50+1=16,944,101 processed words. I doubt that this is faster than having those words cached in a hash.

Please try to cache processed words in a hash so that subsequent `find_anagrams` calls don't have to reprocess the dictionary. You should use the same cache for other methods.